app/api/utils/AppContext.ts

```diff
@@ -0,0 +1,33 @@
+import { AsyncLocalStorage } from 'async_hooks';
+
+interface ContextData {
+  [k: string]: unknown;
+}
+
+class AppContext {
+  private storage = new AsyncLocalStorage<ContextData>();
+
+  async run(cb: () => Promise<void>, data: ContextData = {}): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.storage.run(data, () => {
+        cb()
+          .then(resolve)
+          .catch(reject);
+      });
+    });
+  }
+
+  get(key: string) {
+    const data = this.storage.getStore();
+    return data && data[key];
+  }
+
+  set(key: string, value: unknown) {
+    const data = this.storage.getStore();
+    if (data) data[key] = value;
+  }
```

Comment:
@fnocetti i think we should manage and test this one too.
i feel like there should never be an undefined data ? what scenario apart from context loss can prouce this ? i see this as a potential silent error.

Reply:
see previous. thanks!

Reply:
👍